server/blob/blob-deduplicating/src/main/scala/org/apache/james/server/blob/deduplication/GC.scala

```diff
@@ -0,0 +1,185 @@
+  }
+
+}
+
+/**
+ * NullObject for the initialisation of the GC
+ */
+case object NonExistingGeneration extends Generation {
+  override def previous: Generation = NonExistingGeneration
+
+  override def next: Generation = Generation.first
+
+  override def compareTo(t: Generation): Int = t match {
+    case NonExistingGeneration => 0
+    case _: ValidGeneration => -1
+  }
+}
+
+/**
+ * A run of the GC regarding a Set of Generations
+ */
+case class Iteration(id: Long, processedGenerations: Set[Generation], lastGeneration: Generation) {
+  def next(generations: Set[Generation], lastGeneration: Generation): Iteration = Iteration(id + 1, generations, lastGeneration)
+}
+
+object Iteration {
+  def initial: Iteration = Iteration(0, Set(), NonExistingGeneration)
+}
+
+case class ExternalID(id: String) // TODO
+
+/**
+ * Modelized users' interactions related to blobs
+ */
+sealed trait Event {
+  def blob: BlobId
+  def externalId: ExternalID
+  def generation: Generation
+}
+
+case class Reference(externalId: ExternalID, blobId: BlobId, generation: Generation) extends Event {
+  override def blob: BlobId = blobId
+}
+
+case class Deletion(generation: Generation, reference: Reference) extends Event {
+  override def blob: BlobId = reference.blob
+  override def externalId: ExternalID = reference.externalId
+}
+
+object Events {
+  def getLastGeneration(events: Seq[Event]): Generation = events.map(_.generation).maxOption
+    .getOrElse(Generation.first)
+
+}
+
+case class Report(iteration: Iteration, blobsToDelete: Set[(Generation, BlobId)])
+
+/**
+ * Accessors to the References/Dereferences made by generations
+ */
+case class StabilizedState(references: Map[Generation, Seq[Reference]], deletions: Map[Generation, Seq[Deletion]]) {
+  private val referencedBlobsAcrossGenerations: Map[Generation, ReferencedBlobs] = {
+    val blobIds = references.keys ++ deletions.keys
+    val maxGeneration = blobIds.maxOption.getOrElse(Generation.first)
+    val minGeneration = blobIds.minOption.getOrElse(Generation.first)
+
+    val initialRefs = Generation.range(NonExistingGeneration, minGeneration.previous).map((_, ReferencedBlobs(Map()))).toMap
+    Generation.range(minGeneration, maxGeneration)
+      .foldLeft(initialRefs)(buildGeneration)
+  }
+
+  private def buildGeneration(refs: Map[Generation, ReferencedBlobs], generation: Generation): Map[Generation, ReferencedBlobs] = {
+    val populatedRefs = references.getOrElse(generation, Set())
+      .foldLeft(refs(generation.previous))((currentReferences, ref) => currentReferences.addReferences(ref.blobId))
+
+    val expungedRefs = deletions.getOrElse(generation, Set())
+      .foldLeft(populatedRefs)((currentReferences, ref) => currentReferences.removeReferences(ref.reference.blobId))
+
+    refs + (generation -> expungedRefs)
+  }
+
+  def referencesAt(generation: Generation): ReferencedBlobs = referencedBlobsAcrossGenerations(generation)
+
+  case class ReferencedBlobs(blobs: Map[BlobId, Int]) {
```

Comment:
What does the `Int` part means? Count? Can't we make this obvious by either typing or naming?